In [24]:
#Ejercicio 1: Limpieza de datos inicial
#Visualizacion del csv:

import pandas as pd

ruta = r'df_estados_bank.csv'
df = pd.read_csv(ruta, sep = ',')
df

Unnamed: 0  RowNumber  CustomerId  CreditScore  Gender    Balance  \
0              0          1    15634602          619  Female       0.00   
1              1          2    15647311          608  Female   83807.86   
2              2          3    15619304          502  Female  159660.80   
3              3          4    15701354          699  Female       0.00   
4              4          5    15737888          850  Female  125510.82   
...          ...        ...         ...          ...     ...        ...   
9995        9995       9996    15606229          771    Male       0.00   
9996        9996       9997    15569892          516    Male   57369.61   
9997        9997       9998    15584532          709  Female       0.00   
9998        9998       9999    15682355          772    Male   75075.31   
9999        9999      10000    15628319          792  Female  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember   Age   Geography  \
0                 1          1               1  42.0       Texas   
1                 1          0               1  41.0  California   
2                 3          1               0  42.0       Texas   
3                 2          0               0  39.0       Texas   
4                 1          1               1  43.0  California   
...             ...        ...             ...   ...         ...   
9995              2          1               0  39.0       Texas   
9996              1          1               1  35.0       Texas   
9997              1          0               1   NaN       Texas   
9998              2          1               0  42.0     Alabama   
9999              1          1               0  28.0       Texas   

      EstimatedSalary    Surname  Exited  Tenure  
0           101348.88   Hargrave       1       2  
1           112542.58       Hill       0       1  
2           113931.57       Onio       1       8  
3            93826.63       Boni       0       1  
4            79084.10   Mitchell       0       2  
...               ...        ...     ...     ...  
9995         96270.64   Obijiaku       0       5  
9996        101699.77  Johnstone       0      10  
9997         42085.58        Liu       1       7  
9998         92888.52  Sabbatini       1       3  
9999         38190.78     Walker       0       4  

[10000 rows x 15 columns]

In [25]:
#Viendo la informacion del df:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       10000 non-null  int64  
 1   RowNumber        10000 non-null  int64  
 2   CustomerId       10000 non-null  int64  
 3   CreditScore      10000 non-null  int64  
 4   Gender           10000 non-null  object 
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   Age              9360 non-null   float64
 10  Geography        10000 non-null  object 
 11  EstimatedSalary  9208 non-null   float64
 12  Surname          9706 non-null   object 
 13  Exited           10000 non-null  int64  
 14  Tenure           10000 non-null  int64  
dtypes: float64(3), int64(9), object(3)
memory usage: 1.1+ MB


In [26]:
#LIMPIEZA DE DATOS INICIAL 
missing_values = df.isnull().sum()
missing_values

Unnamed: 0           0
RowNumber            0
CustomerId           0
CreditScore          0
Gender               0
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
Age                640
Geography            0
EstimatedSalary    792
Surname            294
Exited               0
Tenure               0
dtype: int64

*PREGUNTAS PARA DISCUTIR EN EQUIPO:*
¿QUE CONCLUSION PODEMOS SACAR QUE 'AGE' TENGA 640 DATOS NULOS, ASI COMO 'EstimatedSalary' (792) Y 'Surname'(294)?

¿SON IMPORTANTES? O LOS PODEMOS ELIMINAR?

PORCENTAJE DE NULOS EN AGE: 6.4 %
PORCEBTAJE DE NULOS EN ESTIMATED SALARY: 7,92%

¿conviene eliminar aquellos registros que contienen esos pocos valores nulos?
-----> SEGUN LA INFORMACION SE APLICA A VALORES menoRES del 5%

1) PREGUNTA DEL RETO: ¿Consideráis que todas las variables del conjunto de datos son útiles de cara a la predicción? ¿Cuáles sí y cuáles no? 

<RowNumber: Es un índice del registro y no aporta información útil para la predicción. Puede ser eliminado.

<CustomerId: Es un identificador único de cliente y no aporta información predictiva directa. Puede ser eliminado.

<Surname: El apellido del cliente no tiene relevancia en la predicción del churn. Puede ser eliminado.

>CreditScore: La puntuación de crédito del cliente puede ser un predictor importante.

>Geography: El estado del cliente puede influir en su comportamiento bancario.

>Gender: El género del cliente puede tener algún impacto en el comportamiento.

>Age: La edad del cliente es un factor relevante para el análisis.

>Tenure: El número de años que el cliente lleva en el banco es un factor importante.

>Balance: El balance del cliente puede ser crucial para predecir el churn. (AVERIGUAR QUE ES BALANCE)

>NumOfProducts: El número de productos que el cliente tiene en el banco puede influir en su lealtad.

>HasCrCard: Tener o no tarjeta de crédito puede ser relevante.

>IsActiveMember: Ser un cliente activo puede influir en la probabilidad de churn.

>EstimatedSalary: El salario estimado del cliente puede ser un factor importante.

>Exited: Es la variable objetivo que estamos tratando de predecir.

<Por lo tanto, las variables útiles para la predicción son:>

CreditScore
Geography
Gender
Age
Tenure
Balance
NumOfProducts
HasCrCard
IsActiveMember
EstimatedSalary



In [27]:
# Procedemos a eliminar columnas no útiles

df_cleaned = df.drop(columns=['RowNumber', 'CustomerId', 'Surname'])
df_cleaned.head()

Unnamed: 0  CreditScore  Gender    Balance  NumOfProducts  HasCrCard  \
0           0          619  Female       0.00              1          1   
1           1          608  Female   83807.86              1          0   
2           2          502  Female  159660.80              3          1   
3           3          699  Female       0.00              2          0   
4           4          850  Female  125510.82              1          1   

   IsActiveMember   Age   Geography  EstimatedSalary  Exited  Tenure  
0               1  42.0       Texas        101348.88       1       2  
1               1  41.0  California        112542.58       0       1  
2               0  42.0       Texas        113931.57       1       8  
3               0  39.0       Texas         93826.63       0       1  
4               1  43.0  California         79084.10       0       2

----> TRABAJAREMOS CON AGE Y ESTIMATED SALARY, POR LO TANTO IMPUTAMOS:
SUSTITUIMOS LOS "MISSINGS" POR VALORES VALIDOS 

In [29]:
#SUSTITUIMOS LOS "MISSINGS" POR VALORES VALIDOS
#Utilizaremos la función KNNImputer de sklearn.impute

from sklearn.impute import KNNImputer

imputer = KNNImputer()
null_Colums =["Age","EstimatedSalary"]
df_cleaned[null_Colums]=imputer.fit_transform(df_cleaned[null_Colums])

df_cleaned.isnull().sum()

Unnamed: 0         0
CreditScore        0
Gender             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
Age                0
Geography          0
EstimatedSalary    0
Exited             0
Tenure             0
dtype: int64

In [34]:
#Eliminando filas duplicadas
df = df_cleaned.drop_duplicates()
df

# ----> EL DF SE MANTUVO IGUAL, POR LO TANTO, NO HABIAN FILAS DUPLICADAS

Unnamed: 0  CreditScore  Gender    Balance  NumOfProducts  HasCrCard  \
0              0          619  Female       0.00              1          1   
1              1          608  Female   83807.86              1          0   
2              2          502  Female  159660.80              3          1   
3              3          699  Female       0.00              2          0   
4              4          850  Female  125510.82              1          1   
...          ...          ...     ...        ...            ...        ...   
9995        9995          771    Male       0.00              2          1   
9996        9996          516    Male   57369.61              1          1   
9997        9997          709  Female       0.00              1          0   
9998        9998          772    Male   75075.31              2          1   
9999        9999          792  Female  130142.79              1          1   

      IsActiveMember   Age   Geography  EstimatedSalary  Exited  Tenure  
0                  1  42.0       Texas        101348.88       1       2  
1                  1  41.0  California        112542.58       0       1  
2                  0  42.0       Texas        113931.57       1       8  
3                  0  39.0       Texas         93826.63       0       1  
4                  1  43.0  California         79084.10       0       2  
...              ...   ...         ...              ...     ...     ...  
9995               0  39.0       Texas         96270.64       0       5  
9996               1  35.0       Texas        101699.77       0      10  
9997               1  38.4       Texas         42085.58       1       7  
9998               0  42.0     Alabama         92888.52       1       3  
9999               0  28.0       Texas         38190.78       0       4  

[10000 rows x 12 columns]

VER SI ES NECESARIO EL CODIGO DE ABAJO y DESCUBRIR EL PORQUE SE APLICA:

In [35]:
#VERIFICAR, PORQUE LOS DATOS YA ERAN ENTEROS, ¿ERA NECESARIO?
# Convertimos las columnas al tipo de datos adecuado
df['HasCrCard'] = df['HasCrCard'].astype('int')
df['IsActiveMember'] = df['IsActiveMember'].astype('int')
df['Exited'] = df['Exited'].astype('int')

In [36]:
# Información general del DataFrame después de la limpieza
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       10000 non-null  int64  
 1   CreditScore      10000 non-null  int64  
 2   Gender           10000 non-null  object 
 3   Balance          10000 non-null  float64
 4   NumOfProducts    10000 non-null  int64  
 5   HasCrCard        10000 non-null  int32  
 6   IsActiveMember   10000 non-null  int32  
 7   Age              10000 non-null  float64
 8   Geography        10000 non-null  object 
 9   EstimatedSalary  10000 non-null  float64
 10  Exited           10000 non-null  int32  
 11  Tenure           10000 non-null  int64  
dtypes: float64(3), int32(3), int64(4), object(2)
memory usage: 820.4+ KB


COMPARAR CON EL 'df.info()' DEL INCIO

In [37]:
#CONVERSION DE VARIABLES CATEGORICAS A NUMERICAS (GENDER Y GEOGRAPHY)

columnas_categoricas = ['Gender', 'Geography']

# Aplica one-hot encoding
df_encoded = pd.get_dummies(df, columns=columnas_categoricas)

# El nuevo dataframe 'df_encoded' ahora contiene las columnas codificadas
df_encoded.head()


Unnamed: 0  CreditScore    Balance  NumOfProducts  HasCrCard  \
0           0          619       0.00              1          1   
1           1          608   83807.86              1          0   
2           2          502  159660.80              3          1   
3           3          699       0.00              2          0   
4           4          850  125510.82              1          1   

   IsActiveMember   Age  EstimatedSalary  Exited  Tenure  Gender_Female  \
0               1  42.0        101348.88       1       2           True   
1               1  41.0        112542.58       0       1           True   
2               0  42.0        113931.57       1       8           True   
3               0  39.0         93826.63       0       1           True   
4               1  43.0         79084.10       0       2           True   

   Gender_Male  Geography_Alabama  Geography_California  Geography_Texas  
0        False              False                 False             True  
1        False              False                  True            False  
2        False              False                 False             True  
3        False              False                 False             True  
4        False              False                  True            False

In [39]:
#Separamos los datos en conjuntos de entrenamiento y prueba 
#Lo aplicamos al nuevo df (df_encoded)
from sklearn.model_selection import train_test_split

# Definir variables predictoras (X) y objetivo (y)
X = df_encoded.drop(columns=['Exited'])
y = df_encoded['Exited']  # --> Variable Objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

<Estamos enfrentando un problema de *clasificación binaria* (predecir si un cliente abandona o no el banco), por lo que usaremos modelos de clasificación como:

-Regresión Logística
-Árboles de Decisión
-Random Forest
-Gradient Boosting (XGBoost)

In [40]:
#APLICAMOS LOS DIFERENTES ALGORITMOS SELECCIONADOS PARA COMPARAR CUAL ES EL QUE MAS NOS CONVIENE:

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Inicializar los modelos
log_reg = LogisticRegression()
tree_clf = DecisionTreeClassifier()
rf_clf = RandomForestClassifier()
gb_clf = GradientBoostingClassifier()

# Entrenar los modelos
log_reg.fit(X_train, y_train)
tree_clf.fit(X_train, y_train)
rf_clf.fit(X_train, y_train)
gb_clf.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred_log_reg = log_reg.predict(X_test)
y_pred_tree_clf = tree_clf.predict(X_test)
y_pred_rf_clf = rf_clf.predict(X_test)
y_pred_gb_clf = gb_clf.predict(X_test)

#Usamos métricas de evaluación para comparar los modelos:
def evaluate_model(y_test, y_pred):
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    return accuracy, precision, recall, f1

#Evaluamos cada modelo ---------> repasarlo para entenderlo mejor:
metrics_log_reg = evaluate_model(y_test, y_pred_log_reg)
metrics_tree_clf = evaluate_model(y_test, y_pred_tree_clf)
metrics_rf_clf = evaluate_model(y_test, y_pred_rf_clf)
metrics_gb_clf = evaluate_model(y_test, y_pred_gb_clf)

# Resultados
print("Logistic Regression: ", metrics_log_reg)
print("Decision Tree: ", metrics_tree_clf)
print("Random Forest: ", metrics_rf_clf)
print("Gradient Boosting: ", metrics_gb_clf)

C:\Users\maria\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Logistic Regression:  (0.833, 0.4745762711864407, 0.05622489959839357, 0.10053859964093358)
Decision Tree:  (0.816, 0.44886363636363635, 0.4759036144578313, 0.4619883040935672)
Random Forest:  (0.884, 0.7678571428571429, 0.43172690763052207, 0.5526992287917738)
Gradient Boosting:  (0.8863333333333333, 0.7697594501718213, 0.4497991967871486, 0.5678073510773131)


OBSERVACIONES:

-La regresión logística tiene una baja sensibilidad (recall), lo que significa que no identifica bien los casos positivos.
-El árbol de decisión tiene un buen recall, pero su precisión es menor.
-Random Forest y Gradient Boosting tienen un equilibrio entre precisión y recall.

IMPORTANTE: 
Considerar ajustar los hiperparámetros de los modelos (como el número de árboles en Random Forest o la tasa de aprendizaje en Gradient Boosting).
(REVISAR DOCUMENTACION) (REVISAR Y PROBAR CON DISTINTOS PARAMETROS PARA MEJORAR EL MODELO)

In [44]:
#PROBANDOLO NUEVAMENTE, PERO AGREGANDO LOS HIPERPARAMETROS:

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Inicializar los modelos
log_reg = LogisticRegression(max_iter=1000)
tree_clf = DecisionTreeClassifier()
rf_clf = RandomForestClassifier(n_estimators=200, max_depth=10)
gb_clf = GradientBoostingClassifier(learning_rate=0.01, n_estimators=200, max_depth=5)

# Entrenar los modelos
log_reg.fit(X_train, y_train)
tree_clf.fit(X_train, y_train)
rf_clf.fit(X_train, y_train)
gb_clf.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred_log_reg = log_reg.predict(X_test)
y_pred_tree_clf = tree_clf.predict(X_test)
y_pred_rf_clf = rf_clf.predict(X_test)
y_pred_gb_clf = gb_clf.predict(X_test)

#Usamos métricas de evaluación para comparar los modelos:
def evaluate_model(y_test, y_pred):
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    return accuracy, precision, recall, f1

#Evaluamos cada modelo ---------> repasarlo para entenderlo mejor:
metrics_log_reg = evaluate_model(y_test, y_pred_log_reg)
metrics_tree_clf = evaluate_model(y_test, y_pred_tree_clf)
metrics_rf_clf = evaluate_model(y_test, y_pred_rf_clf)
metrics_gb_clf = evaluate_model(y_test, y_pred_gb_clf)

# Resultados
print("Logistic Regression: ", metrics_log_reg)
print("Decision Tree: ", metrics_tree_clf)
print("Random Forest: ", metrics_rf_clf)
print("Gradient Boosting: ", metrics_gb_clf)

C:\Users\maria\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Logistic Regression:  (0.8433333333333334, 0.5777777777777777, 0.20883534136546184, 0.30678466076696165)
Decision Tree:  (0.819, 0.45596868884540115, 0.4678714859437751, 0.4618434093161546)
Random Forest:  (0.8813333333333333, 0.77734375, 0.39959839357429716, 0.5278514588859416)
Gradient Boosting:  (0.883, 0.8356164383561644, 0.3674698795180723, 0.5104602510460251)


In [ ]:
#Selección del mejor modelo
#Comparar las métricas obtenidas para seleccionar el modelo con mejor rendimiento

***MEJOR MODELO****
El modelo con mayor rendimiento en términos de precisión y F1-score es el "Random Forest". Aunque su recall es más bajo, su equilibrio entre precisión y recall lo hace destacar.

EJERICIO 2: 
¿QUE ALGORITMO ES EL OS HA ARROJADO MEJORES RESULTADOS?
-RANDOM FOREST
¿COMO HABEIS EVALUADO LA BONDAD DEL AJUSTE?
-DE ACUERDO A LAS METRICAS (NOMBRARLAS)

*SIGUE* ---->

In [45]:
# Análisis de características importantes
#Identificar y explicar las características más influyentes en la decisión de un cliente para abandonar la entidad utilizando la importancia de características de modelos como Random Forest
importances_rf = rf_clf.feature_importances_

# Mostrar la importancia de características
print("Random Forest Feature Importances: ", importances_rf)

Random Forest Feature Importances:  [0.06814878 0.06893085 0.09575254 0.17609707 0.00980273 0.04534375
 0.24981787 0.06983681 0.03783182 0.01111232 0.0113117  0.05913316
 0.07022853 0.02665209]


---> ¿PODEIS EXPLICAR QUE CARACTERISTICAS SON LAS QUE MAS INCIDEN EN LA DECISION DE UN CLIENTE PARA ABANDONAR DEFINITIVAMENTE LA ENTIDAD?

- Características más importantes:
La característica con el índice 6 (EstimatedSalary) tiene la mayor importancia, con un valor de aproximadamente 0.25.
La característica con el índice 3 (HasCrCard) también es relevante, con un valor de alrededor de 0.18.

In [50]:
#¿Obtenéis mejores resultados planteando modelos diferenciados por estado?

states = ['Alabama', 'California', 'Texas']
results_by_state = {}

for state in states:
    state_data = df_encoded[df_encoded['Geography_' + state] == 1]
    X_state = state_data.drop(columns=['Exited', 'Geography_Alabama', 'Geography_California', 'Geography_Texas'])
    y_state = state_data['Exited']
    X_train_state, X_test_state, y_train_state, y_test_state = train_test_split(X_state, y_state, test_size=0.3, random_state=42)
    
    rf_clf_state = RandomForestClassifier(n_estimators=200, max_depth=10)
    rf_clf_state.fit(X_train_state, y_train_state)
    
    y_pred_state = rf_clf_state.predict(X_test_state)
    results_by_state[state] = evaluate_model(y_test_state, y_pred_state)

# Mostrar resultados por estado
results_by_state

C:\Users\maria\AppData\Roaming\Python\Python311\site-packages\sklearn\metrics\_classification.py:1517: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


{'Alabama': (0.798140770252324,
  0.8156424581005587,
  0.5509433962264151,
  0.6576576576576577),
 'California': (0.9623655913978495, 0.0, 0.0, 0.0),
 'Texas': (0.8764119601328904, 0.8425925925925926, 0.35, 0.4945652173913043)}

INTERPRETAR LOS RESULTADOS DE ARRIBA.

 TENER EN CUENTA QUE AUN QUEDA UNA PREGUNTA: ¿Tenéis algún problema de desbalanceo de datos? Si es así, solucionadlo.

 DESBALANCEO DE DATOS: COMO LO SABEMOS?

In [54]:
# EJERCICIO 4: Algoritmo de clusterización
#Aplicando K-means para encontrar grupos homogéneos de clientes y describir las características de cada grupo

from sklearn.cluster import KMeans
import numpy as np

# Preparar los datos para la clusterización
X_cluster = df_encoded.drop(columns=['Exited'])

# Aplicar K-means
kmeans = KMeans(n_clusters=4, random_state=42)
kmeans.fit(X_cluster)

# Asignar etiquetas de cluster
df_encoded['Cluster'] = kmeans.labels_

# Describir características de cada grupo
clusters = df_encoded.groupby('Cluster').mean()

clusters

Unnamed: 0  CreditScore        Balance  NumOfProducts  HasCrCard  \
Cluster                                                                      
0        5026.826602   648.516424    2686.347943       1.767905   0.711362   
1        5016.709661   650.175745    2234.948243       1.764132   0.725591   
2        5005.449691   651.503419  121881.158372       1.371866   0.700098   
3        4966.708253   650.986564  121952.686536       1.398912   0.694818   

         IsActiveMember        Age  EstimatedSalary    Exited    Tenure  \
Cluster                                                                   
0              0.518578  38.552073    148517.496775  0.109855  7.299946   
1              0.517986  38.513052     53084.732996  0.103803  6.997431   
2              0.514165  39.404429     53174.964544  0.206773  7.492999   
3              0.512156  38.847217    147967.005380  0.211772  7.035829   

         Gender_Female  Gender_Male  Geography_Alabama  Geography_California  \
Cluster                                                                        
0             0.464728     0.535272           0.016155              0.328487   
1             0.452724     0.547276           0.013361              0.323741   
2             0.440247     0.559753           0.393032              0.199284   
3             0.462892     0.537108           0.398592              0.199936   

         Geography_Texas  
Cluster                   
0               0.655358  
1               0.662898  
2               0.407685  
3               0.401472

REALIZAR GRAFICO PARA VISUALIZARLOS :



¿Tiene sentido dividir los datos en 4 grupos? ¿O tal vez deberíamos explorar más o menos clusters?

REVISAR DOCUMENTACION, HABIA UN MODO DE HALLAR LA CANTIDAD ADECUADA POR MEDIO DE UN GRAFICO (ALGO DEL CODO)